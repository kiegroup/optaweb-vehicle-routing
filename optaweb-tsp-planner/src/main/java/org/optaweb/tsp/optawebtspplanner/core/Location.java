package org.optaweb.tsp.optawebtspplanner.core;

import java.util.Objects;

/**
 * A unique location significant to the user.
 */
public class Location {

    private final long id;
    private final LatLng latLng;

    public Location(long id, LatLng latLng) {
        Objects.requireNonNull(latLng);
        this.id = id;
        this.latLng = latLng;
    }

    public long getId() {
        return id;
    }

    public LatLng getLatLng() {
        return latLng;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Location that = (Location) o;
        return id == that.id &&
                latLng.equals(that.latLng);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, latLng);
    }

    @Override
    public String toString() {
        return "Location{" +
                "id=" + id +
                ", latLng=" + latLng +
                '}';
    }
}
