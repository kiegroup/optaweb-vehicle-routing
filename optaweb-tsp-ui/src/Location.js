import PropTypes from 'prop-types';
import React from 'react';

function Location({ id, removeDisabled, removeHandler, selectHandler }) {
  return (
    <div
      key={id}
      className={'ma2 flex bg-animate hover-bg-light-gray'}
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <span className={'w-80 pa2'}>{`Location ${id}`}</span>
      <button disabled={removeDisabled} className={'w-20 pa2'} onClick={() => removeHandler(id)}>x
      </button>
    </div>
  );
}

Location.propTypes = {
  id: PropTypes.number.isRequired,
  removeDisabled: PropTypes.bool.isRequired,
  removeHandler: PropTypes.func.isRequired,
  selectHandler: PropTypes.func.isRequired,
};

export default Location;
