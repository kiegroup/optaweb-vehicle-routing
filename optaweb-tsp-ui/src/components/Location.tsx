/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

export interface ILocationProps {
  id: number;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (e: any) => void; // FIXME: Event Type
}

const Location: React.SFC<ILocationProps> = ({
  id,
  removeDisabled,
  removeHandler,
  selectHandler
}: ILocationProps) => {
  return (
    <div
      key={id}
      className='ma2 flex bg-animate hover-bg-light-gray'
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <span className='w-80 pa2'>{`Location ${id}`}</span>
      <button
        type='button'
        disabled={removeDisabled}
        className='w-20 pa2'
        onClick={() => removeHandler(id)}
      >
        x
      </button>
    </div>
  );
};

export default Location;
