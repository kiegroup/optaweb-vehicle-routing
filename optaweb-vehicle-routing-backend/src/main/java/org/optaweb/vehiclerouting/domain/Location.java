/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.vehiclerouting.domain;

import java.util.Objects;

/**
 * A unique location significant to the user.
 */
public class Location extends LocationData {

    private final long id;

    public Location(long id, Coordinates coordinates) {
        // TODO remove this?
        this(id, coordinates, "");
    }

    public Location(long id, Coordinates coordinates, String description) {
        super(coordinates, description);
        this.id = id;
    }

    /**
     * Location's ID.
     * @return unique ID
     */
    public long id() {
        return id;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Location location = (Location) o;
        return id == location.id;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), id);
    }

    @Override
    public String toString() {
        return "Location ["
                + id +
                "]: '" + description() + "'";
    }
}
