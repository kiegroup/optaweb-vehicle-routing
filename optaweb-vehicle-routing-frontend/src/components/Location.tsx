/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Grid, GridItem } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import * as React from 'react';

export interface ILocationProps {
  id: number;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
}

const Location: React.SFC<ILocationProps> = ({
  id,
  removeDisabled,
  removeHandler,
  selectHandler,
}: ILocationProps) => {
  return (
    <Grid
      gutter="md"
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <GridItem span={9}>
        {`Location ${id}`}
      </GridItem>
      <GridItem span={3}>
        <Button
          variant="link"
          isDisabled={removeDisabled}
          onClick={() => removeHandler(id)}
          type="button"
        >
          <TimesIcon />
        </Button>
      </GridItem>
    </Grid>
  );
};

export default Location;
