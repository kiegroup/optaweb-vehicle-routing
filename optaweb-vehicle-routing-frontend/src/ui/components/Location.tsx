/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, DataListCell, DataListItem, DataListItemRow, Tooltip } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import * as React from 'react';

export interface LocationProps {
  id: number;
  description: string | null;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
}

const Location: React.FC<LocationProps> = ({
  id,
  description,
  removeDisabled,
  removeHandler,
  selectHandler,
}) => {
  const [clicked, setClicked] = React.useState(false);

  function shorten(text: string) {
    const first = text.replace(/,.*/, '').trim();
    const short = first.substring(0, Math.min(20, first.length)).trim();
    if (short.length < first.length) {
      return `${short}...`;
    }
    return short;
  }

  return (
    <DataListItem
      isExpanded={false}
      aria-labelledby={`location-${id}`}
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <DataListItemRow>
        <DataListCell isFilled>
          {(description && (
            <Tooltip content={description}>
              <span id={`location-${id}`}>{shorten(description)}</span>
            </Tooltip>
          ))
          || <span id={`location-${id}`}>{`Location ${id}`}</span>}
        </DataListCell>
        <DataListCell isFilled={false}>
          <Button
            type="button"
            variant="link"
            isDisabled={removeDisabled || clicked}
            onClick={() => {
              setClicked(true);
              removeHandler(id);
            }}
          >
            <TimesIcon />
          </Button>
        </DataListCell>
      </DataListItemRow>
    </DataListItem>
  );
};

export default Location;
