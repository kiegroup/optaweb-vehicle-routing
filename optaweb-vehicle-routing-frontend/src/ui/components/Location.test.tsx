import { Button, DataListItem } from '@patternfly/react-core';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import Location, { LocationProps } from './Location';

describe('Location Component', () => {
  it('should render correctly', () => {
    const props: LocationProps = {
      id: 10,
      description: 'x',
      removeDisabled: false,
      removeHandler: jest.fn(),
      selectHandler: jest.fn(),
    };
    const location = shallow(<Location {...props} />);
    expect(toJson(location)).toMatchSnapshot();
    location.find(DataListItem).simulate('mouseEnter');
    location.find(Button).simulate('click');

    expect(props.removeHandler).toHaveBeenCalledTimes(1);
    expect(props.selectHandler).toHaveBeenCalledTimes(1);
  });

  it('should render correctly when description is missing', () => {
    const props: LocationProps = {
      id: 11,
      description: null,
      removeDisabled: false,
      removeHandler: jest.fn(),
      selectHandler: jest.fn(),
    };
    const location = shallow(<Location {...props} />);
    expect(toJson(location)).toMatchSnapshot();
  });
});
