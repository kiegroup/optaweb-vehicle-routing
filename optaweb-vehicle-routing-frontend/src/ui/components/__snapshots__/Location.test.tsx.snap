// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location Component should render correctly 1`] = `
<DataListItem
  aria-labelledby="location-10"
  className=""
  isExpanded={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <DataListItemRow
    className=""
  >
    <DataListCell
      alignRight={false}
      className=""
      isFilled={true}
      isIcon={false}
      width={1}
    >
      <Tooltip
        appendTo={[Function]}
        className={null}
        content="x"
        enableFlip={true}
        entryDelay={500}
        exitDelay={500}
        maxWidth="18.75rem"
        position="top"
        trigger="mouseenter focus"
        zIndex={9999}
      >
        <span
          id="location-10"
        >
          x
        </span>
      </Tooltip>
    </DataListCell>
    <DataListCell
      alignRight={false}
      className=""
      isFilled={false}
      isIcon={false}
      width={1}
    >
      <Button
        isDisabled={false}
        onClick={[Function]}
        type="button"
        variant="link"
      >
        <TimesIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </DataListCell>
  </DataListItemRow>
</DataListItem>
`;

exports[`Location Component should render correctly when description is missing 1`] = `
<DataListItem
  aria-labelledby="location-11"
  className=""
  isExpanded={false}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <DataListItemRow
    className=""
  >
    <DataListCell
      alignRight={false}
      className=""
      isFilled={true}
      isIcon={false}
      width={1}
    >
      <span
        id="location-11"
      >
        Location 11
      </span>
    </DataListCell>
    <DataListCell
      alignRight={false}
      className=""
      isFilled={false}
      isIcon={false}
      width={1}
    >
      <Button
        isDisabled={false}
        onClick={[Function]}
        type="button"
        variant="link"
      >
        <TimesIcon
          color="currentColor"
          size="sm"
          title={null}
        />
      </Button>
    </DataListCell>
  </DataListItemRow>
</DataListItem>
`;
